pkg/job/scraper.go

```diff
@@ -0,0 +1,240 @@
+
+					jobToRun = cloudwatchrunner.DiscoveryJob{Job: job, Resources: resources}
+				}, func(job model.CustomNamespaceJob) {
+					jobToRun = cloudwatchrunner.CustomNamespaceJob{Job: job}
+				},
+			)
+			if jobToRun == nil {
+				jobLogger.Debug("Ending job run early due to job error see job errors")
+				return
+			}
+			jobLogger.Debug("Starting cloudwatch metrics runner")
+			cwRunner := s.runnerFactory.NewCloudWatchRunner(jobLogger, region, role, jobToRun)
+			metricResult, err := cwRunner.Run(ctx)
+			if err != nil {
+				jobError := Error{
+					AccountID: accountID,
+					Region:    region,
+					RoleARN:   role.RoleArn,
+					Namespace: namespace,
+					Message:   "Failed to gather cloudwatch metrics for job",
+					Err:       err,
+				}
+				mux.Lock()
+				jobErrors = append(jobErrors, jobError)
+				mux.Unlock()
+
+				return
+			}
+
+			if len(metricResult) == 0 {
+				jobLogger.Debug("No metrics data found")
+				return
+			}
+
+			jobLogger.Debug("Job run finished", "number_of_metrics", len(metricResult))
+
+			result := model.CloudwatchMetricResult{
+				Context: &model.ScrapeContext{
+					Region:     region,
+					AccountID:  accountID,
+					CustomTags: jobToRun.CustomTags(),
+				},
+				Data: metricResult,
+			}
+
+			mux.Lock()
+			defer mux.Unlock()
+			metricResults = append(metricResults, result)
+		}()
+	})
+	wg.Wait()
+	s.logger.Debug("Finished job runs", "resource_results", len(resourceResults), "metric_results", len(metricResults))
+	return resourceResults, metricResults, jobErrors
+}
+
+// Walk through each custom namespace and discovery jobs and take an action
+func jobConfigVisitor(jobsCfg model.JobsConfig, action func(job any, role model.Role, region string)) {
+	for _, job := range jobsCfg.DiscoveryJobs {
+		for _, role := range job.Roles {
+			for _, region := range job.Regions {
+				action(job, role, region)
+			}
+		}
+	}
+
+	for _, job := range jobsCfg.CustomNamespaceJobs {
+		for _, role := range job.Roles {
+			for _, region := range job.Regions {
+				action(job, role, region)
+			}
+		}
+	}
+}
+
+// Take an action depending on the job type, only supports discovery and custom job types
+func jobAction(logger logging.Logger, job any, discovery func(job model.DiscoveryJob), custom func(job model.CustomNamespaceJob)) {
+	// Type switches are free https://stackoverflow.com/a/28027945
+	switch typedJob := job.(type) {
+	case model.DiscoveryJob:
+		discovery(typedJob)
+	case model.CustomNamespaceJob:
+		custom(typedJob)
+	default:
+		logger.Error(fmt.Errorf("config type of %T is not supported", typedJob), "Unexpected job type")
+		return
+	}
+}
+
+type Error struct {
+	AccountID string
+	Namespace string
+	Region    string
+	RoleARN   string
+	Message   string
+	Err       error
+}
+
+func (e Error) ToLoggerKeyVals() []interface{} {
+	return []interface{}{
+		"account_id", e.AccountID,
+		"namespace", e.Namespace,
+		"region", e.Region,
+		"role_arn", e.RoleARN,
+	}
+}
```

Comment:
Rather customer error type to ensure we don't miss contextual logging values we previously had when error logging happened during scraping.